new(config-babel): allow override of ignore

williaster opened the pull request:
We have an issue where we need to allow babel-parsing of one or more packages from `node_modules/`, but we cannot currently override this because `ignore` cannot be affected using babel `overrides` or `exclude`.

This adds a hook to customize the `ignore` list but shouldn't change existing behavior.

@etr2460 @kristw @hayes

williaster committed:
new(config-babel): allow override of ignore

kristw commented:
great. I had to opt-out and define full babel.config.js in the past for this.

etr2460 reviewed (approved):
makes sense to me I think. how did you test this though?

Also, i thought you could already pass in an `ignore` option and it would append it to the outputted array. Where did that functionality live?

kristw commented:
> Also, i thought you could already pass in an ignore option and it would append it to the outputted array. Where did that functionality live?

It will append to the default ignore list, but cannot overwrite. For example, `node_modules` is in the default list. My use case was that I want to remove it and add `node_modules/(?!some-lib)` instead, to parse some modules that export as es modules. However, what ended up happen in the current one is

`['node_modules', 'node_modules/(?!some-lib)']`

which means my added config is useless.

milesj commented:
This still won't change what you expect because the actual config is here: https://github.com/airbnb/nimbus/blob/master/packages/nimbus/src/configs/babel.ts

williaster commented:
yeah in testing this it doesn't work. I also tried updating the file that @milesj linked to extract `babel.ignore` from the `package.nimbus.babel` config, but it would be a breaking change since previously it was merged and now it is used as the base. we could also add a `baseBabelIgnore` option defined at the `package.nimbus.settings` level, but that doesn't seem super clean 😢 

after poking around, since the merging is handled at the [beemo `Driver` layer](https://github.com/beemojs/beemo/blob/068e0c638ee3831c93229e3ac143431f08f83bed/packages/core/src/Driver.ts#L76), I'm not sure of a super clean way to allow this without using `strategy: 'none'` for the babel driver and committing the file (but even then we have to handle `dev` vs `prod` vs `test` config which isn't great either).

I was able to solve our immediate problem with this hack but open to other thoughts 😱 

```ts
const defaultBabelConfig = require('@airbnb/nimbus/configs/babel');

const defaultIgnoreWithoutNodeModules = defaultBabelConfig.ignore.filter(
  pattern => !pattern.includes('node_modules'),
);

// this allows us to override the base ignore array for the driver, rather than merge into it
delete defaultBabelConfig.ignore;

module.exports = {
  ...restConfig
  // currently XXX is built as a module which needs to be parsed by
  // babel-jest else it will throw. babel-jest throws if babel itself doesn't parse it
  // so we ignore all node_modules/ except XXX
  ignore: ['node_modules/(?!XXX)/', ...defaultIgnoreWithoutNodeModules],
};

```

milesj commented:
Yeah, this is also a problem in Beemo itself. I'm not sure how to support merging and overrides in parallel.

williaster commented:
Given all of the above, gonna close this one

milesj commented:
I'm working on beemo v2. I'll try and figure this out as part of that.

kristw commented:
Could the `ignore` parameter type be something like

```ts
type Ignore = string[] | { merge: string[] } | { override: string[] }
```

passing `string[]` is the same with `{ merge: string[] }`, making it backward-compatible.
if `{ override: string[] }` will override and replace with the specified input.

Just throwing in ideas.

milesj commented:
Yeah my current thoughts are similar to that, but using an import instead. Something like:

```
import { override } from '@beemo/core';

module.exports = {
  ignore: override(['new/paths']),
};
```

kristw commented:
> Yeah my current thoughts are similar to that, but using an import instead. Something like:

This looks great but then we can't customize this from `package.json` via `nimbus`? Would be nice to have a `JSON` compatible solution.

milesj commented:
`package.json` configuration has been removed entirely from beemo v2.
